timemachine/fe/functional.py

```diff
@@ -13,38 +16,64 @@ def _make_selection_mask(compute_du_dx=False, compute_du_dp=False, compute_du_dl
 
 
 def wrap_impl(impl, pack=lambda x: x):
-    """Construct a differentiable function U(x, params, box, lam) -> float
+    """Construct a differentiable function U(coords, params, box, lam) -> float
     from a single unbound potential
     """
 
     @custom_jvp
-    def U(coords, params, box, lam):
+    def U(coords, params, box, lam) -> float:
         selection = _make_selection_mask(compute_u=True)
         result_tuple = impl.execute_selective(coords, pack(params), box, lam, *selection)
         return result_tuple[3]
 
-    def U_jvp_x(coords_dot, _, coords, params, box, lam):
-        selection = _make_selection_mask(compute_du_dx=True)
-        result_tuple = impl.execute_selective(coords, pack(params), box, lam, *selection)
-        return np.sum(coords_dot * result_tuple[0])
+    @U.defjvp
+    def U_jvp(primals, tangents) -> Tuple[float, float]:
 
-    def U_jvp_params(params_dot, _, coords, params, box, lam):
-        selection = _make_selection_mask(compute_du_dp=True)
-        result_tuple = impl.execute_selective(coords, pack(params), box, lam, *selection)
-        return np.sum(pack(params_dot) * result_tuple[1])
+        # unpack inputs
+        coords, _params, box, lam = primals
+        coords_t, _params_t, box_t, lam_t = tangents
 
-    def U_jvp_lam(lam_dot, _, coords, params, box, lam):
-        selection = _make_selection_mask(compute_du_dl=True)
-        result_tuple = impl.execute_selective(coords, pack(params), box, lam, *selection)
-        return np.sum(lam_dot * result_tuple[2])
+        # handle case where _params is a list of (maybe traced) arrays
+        params, params_t = pack(_params), pack(_params_t)
+
+        # inspect tangent types to determine which derivatives are being requested
+        def derivative_requested(array_t):
+            return isinstance(array_t, Tracer)
+
+        selection = _make_selection_mask(
+            compute_du_dx=derivative_requested(coords_t),
+            compute_du_dp=derivative_requested(params_t),
+            compute_du_dl=derivative_requested(lam_t),
+            compute_u=True,
+        )
+
+        if derivative_requested(box_t):
+            raise RuntimeError("box derivatives not supported!")
+
+        # call custom op once
+        result_tuple = impl.execute_selective(coords, params, box, lam, *selection)
+
+        # unpack result tuple
+        primal_out = result_tuple[3]
+        coords_g, params_g, lam_g = result_tuple[:3]
+
+        # compute tangent_out by dotting tangents with jacobians
+        # TODO: make this a zippy comprehension?
+        tangent_out = 0.0
+        if derivative_requested(coords_t):
+            tangent_out += np.sum(coords_t * coords_g)
```

Comment:
nit: this should implement a jvp (matrix-vector product). As written, this looks like a `vjp`. 

The `J` here should be `coords_g`, which has shape `(1,3N)`, and `v=coords_t` has shape `(3N,)`. So I think you can simply rewrite this as `tangent_out += coords_g @ coords_t`

Reply:
Will look into this -- I assumed writing as `np.sum(coords_g * coords_t)` was fine since `U` is scalar-valued.

>So I think you can simply rewrite this as tangent_out += coords_g @ coords_t

Will look into this further -- replacing with `coords_g @ coords_t` locally introduces errors like `TypeError: dot_general requires contracting dimensions to have the same shape, got (3,) and (35,).`

Reply:
Note that `coords_g` and `coords_t` currently have the same shape `(n_atoms, 3)`.

I tried a few ways to expand `coords_g` (n_atoms, 3) to `coords_J` of shape `(1, n_atoms, 3)` and replace elementwise multiplication with matrix multiplication, but didn't view any of these as a safety or readability improvement yet...

Unless you have a strong preference here, I'd prefer to keep the implementation as-is, but just clarify naming and add a note that this is specialized to the case of `U` being scalar-valued, as in https://github.com/proteneer/timemachine/pull/707/commits/7bc5f6b709bd2a44e0ff988ef2eb39610145d53b ?

Reply:
Clarified in https://github.com/proteneer/timemachine/pull/707/commits/fb584904771cd22b2282bc5ed7e817d58319573b !